simSPI/crd.py

```diff
@@ -0,0 +1,135 @@
+    numpart : int
+        Number of particles.
+    xrange : ndarray
+        Valid horizontal range for display.
+    yrange : ndarray
+        Valid vertical range for display.
+    crd_file : str
+        Relative path to output .crd file.
+    """
+    if os.path.exists(crd_file):
+        print(crd_file + " already exists.")
+    else:
+        rotlist = get_rotlist(numpart)
+        with open(crd_file, "w") as crd:
+            crd.write("# File created by TEM-simulator, version 1.3.\n")
+            crd.write("{numpart}  6\n".format(numpart=numpart))
+            crd.write(
+                "#            \
+                x             \
+                y             \
+                z             \
+                phi           \
+                theta         \
+                psi  \n"
+            )
+            i = 0
+            for y in yrange:
+                for x in xrange:
+                    if i == int(numpart):
+                        break
+                    crd_table = {
+                        "x": x,
+                        "y": y,
+                        "z": 0,
+                        "phi": rotlist[i][0],
+                        "theta": rotlist[i][1],
+                        "psi": rotlist[i][2],
+                    }
+                    crd.write(
+                        "{0[x]:14.4f} \
+                        {0[y]:14.4f} \
+                        {0[z]:14.4f} \
+                        {0[phi]:14.4f} \
+                        {0[theta]:14.4f} \
+                        {0[psi]:14.4f}\n".format(
+                            crd_table
+                        )
+                    )
+                    i += 1
+
+
+def get_rotlist(numpart):
+    """Return a rotation list containing Euler angles.
+
+    Parameters
+    ----------
+    numpart : int
+        Number of particles.
+    """
+    rotlist = []
+    for x in range(0, numpart + 1):
+        x = special_ortho_group.rvs(3)
+        y = rotation_matrix_to_euler_angles(x)
+        rotlist.append(y)
+    return rotlist
+
+
+def rotation_matrix_to_euler_angles(R):
+    """Compute Euler angles given a rotation matrix.
+
+    Parameters
+    ----------
+    R : ndarray
+        Rotation matrix.
+    """
+    if not is_rotation_matrix(R):
+        raise ValueError()
+
+    sy = math.sqrt(R[0, 0] * R[0, 0] + R[1, 0] * R[1, 0])
+    singular = sy < 1e-6
+    if not singular:
+        x = math.atan2(R[2, 1], R[2, 2])
+        y = math.atan2(-R[2, 0], sy)
+        z = math.atan2(R[1, 0], R[0, 0])
+    else:
+        x = math.atan2(-R[1, 2], R[1, 1])
+        y = math.atan2(-R[2, 0], sy)
+        z = 0
+    x = rad2deg(x)
+    y = rad2deg(y)
+    z = rad2deg(z)
+    return np.array([x, y, z])
+
+
+def rad2deg(x):
+    """Take a value x in radians and convert to degrees.
+
+    Parameters
+    ----------
+    x : float
+        Degree value to convert to degrees.
+    """
+    return (x * 180) / np.pi
+
+
+def is_rotation_matrix(matrix):
```

Comment:
Might be simpler to just check that the determinant of the matrix is 1?
https://numpy.org/doc/stable/reference/generated/numpy.linalg.det.html

Reply:
That's necessary, but not sufficient. we also need R=R.T. For instance [[a,a-1],[a+1,a]] has a det of 1 = a^2 - (a^2 -1), but is not a valid rotation matrix because it's not symmetric.

Now that I think of it, you should also check the determinant is one, not just that R=R.T, which is what the code does below.

```
is_symmetric = np.allclose(matrix,matrix.T)
is_det_one = np.isclose(np.linalg.det(matrix),1)
return is_symmetric and is_det_one
```

Reply:
Hi @geoffwoollard, just wanted to check whether the orthogonality condition for rotation matrices should be R • R.T = I instead of R = R.T.

Reply:
Actually, I was wrong. a 3D rotation matrix is not in general symmetric!!!

Sorry XD

I'm not sure what all the conditions are called exactly, but we should definitely have
https://en.wikipedia.org/wiki/3D_rotation_group#Orthogonal_and_rotation_matrices

R dot Rt = Identity
(det R)**2 = 1

Reply:
No worries! Thanks for clarifying.